ext/drreg/drreg.c

```diff
@@ -1599,18 +2385,52 @@ drreg_restore_app_aflags(void *drcontext, instrlist_t *ilist, instr_t *where)
  */
 
 static bool
-is_our_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT,
-                        reg_id_t *reg_spilled OUT, uint *slot_out OUT, uint *offs_out OUT)
+is_our_spill_or_restore(void *drcontext, instr_t *instr, instr_t *next_instr,
+                        bool *spill OUT, reg_id_t *reg_spilled OUT, uint *slot_out OUT,
+                        uint *offs_out OUT, bool *is_indirectly_spilled OUT)
 {
     bool tls;
     uint slot, offs;
     reg_id_t reg;
-    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, spill, &reg, &offs))
+    bool is_spilled;  /* Flag denoting spill or restore*/
+    bool is_indirect; /* Flag denoting direct or indirect access */
+
+    is_indirect = false;
+    if (is_indirectly_spilled != NULL)
+        *is_indirectly_spilled = is_indirect;
+
+    if (!instr_is_reg_spill_or_restore(drcontext, instr, &tls, &is_spilled, &reg, &offs))
         return false;
-    /* Is this from our raw TLS? */
+    /* Is this from our direct raw TLS for gpr registers ? */
     if (tls && offs >= tls_slot_offs &&
         offs < (tls_slot_offs + ops.num_spill_slots * sizeof(reg_t))) {
         slot = (offs - tls_slot_offs) / sizeof(reg_t);
+    } else if (tls && offs == tls_simd_offs &&
+               !(is_spilled) /* Cant be a spill bc loading block */) {
+        ASSERT(next_instr != NULL, "next_instr cannot be NULL");
+        ASSERT(instr_get_opcode(next_instr) == OP_movdqa ||
+                   instr_get_opcode(next_instr) == OP_vmovdqa,
+               "next instruction needs to be a mov");
+        is_indirect = true;
+        opnd_t dst = instr_get_dst(next_instr, 0);
+        opnd_t src = instr_get_src(next_instr, 0);
+
+        if (opnd_is_reg(dst) && reg_is_vector_simd(opnd_get_reg(dst)) &&
```

Comment:
Is this enough to determine that this is _our_ spill/restore? Could be a tool's instruction.

Reply:
I guess this relies on that the simd mov is right after the indirect block load. I will add a comment.

Reply:
Done. This is a limitation with regards to this approach: If the user wishes to sequentially reserve multiple SIMD registers at once, drreg is required to load the pointer to the indirect block for every spill/restore even though one load at the beginning would suffice.

Reply:
Sure, please do not edit the patch as long as I am working on it as discussed. Otherwise it will be impossible to share the work.

Reply:
Yes, I am re-factoring the entire patch w.r.t. ARM for example. I can push an update in a few hours but my understanding was that you currently can't work on this. Unfortunately we need this functionality here now so I am planning to work on it until you are free.

Reply:
Ah I see, I thought you were addressing the concerns you highlighted in the review instead of a whole refactor, hence enabling me to continue incrementally contributing.. Okay, no problem, thanks for your time.

Reply:
I will keep you updated, let me now should you become more free and want to add more work, then we can sync and update the patch here, no prob.

Reply:
Re: commit 06bcbe54a5dd02e7132ac570ee1ced262d082c68
I am not done yet.

Reply:
> drreg is required to load the pointer to the indirect block for every spill/restore even though one load at the beginning would suffice

I don't see a better way w/o having https://github.com/DynamoRIO/dynamorio/issues/3801.